src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -30,42 +30,111 @@ namespace ArborX
 namespace Details
 {
 
-struct WeightedEdge
+class WeightedEdge
 {
-  int source;
-  int target;
-  float weight;
+public:
+  unsigned long long _directed_edge = ULLONG_MAX;
+  static constexpr auto inf =
+      KokkosExt::ArithmeticTraits::infinity<float>::value;
+  float _weight = inf;
+
+  static_assert(sizeof(unsigned long long) == 8, "");
+  static constexpr int source_shift = 32;
+  static constexpr int target_shift = 1;
+  static constexpr unsigned long long mask_source =
+      static_cast<unsigned long long>(UINT_MAX >> 1) << source_shift;
+  static constexpr unsigned long long mask_target =
+      static_cast<unsigned long long>(UINT_MAX >> 1) << target_shift;
+  static constexpr unsigned long long mask_undirected = ~((1ull << 63) + 1);
+  static_assert(mask_undirected == mask_source + mask_target, "");
+  static constexpr unsigned long long reverse_direction = 1;
 
-private:
   // performs lexicographical comparison by comparing first the weights and then
   // the unordered pair of vertices
   friend KOKKOS_FUNCTION constexpr bool operator<(WeightedEdge const &lhs,
                                                   WeightedEdge const &rhs)
   {
-    if (lhs.weight != rhs.weight)
-    {
-      return (lhs.weight < rhs.weight);
-    }
-    using KokkosExt::min;
-    auto const lhs_min = min(lhs.source, lhs.target);
-    auto const rhs_min = min(rhs.source, rhs.target);
-    if (lhs_min != rhs_min)
+    if (lhs._weight != rhs._weight)
     {
-      return (lhs_min < rhs_min);
+      return (lhs._weight < rhs._weight);
     }
-    using KokkosExt::max;
-    auto const lhs_max = max(lhs.source, lhs.target);
-    auto const rhs_max = max(rhs.source, rhs.target);
-    return (lhs_max < rhs_max);
+    // return ((lhs._directed_edge & mask_undirected) <
+    //         (rhs._directed_edge & mask_undirected));
+    return (lhs._directed_edge < rhs._directed_edge);
```

Comment:
Note that with the current implementation, `(a, b) < (b,a)` for `a > b`. So, same edge will not compare equal, if `a == b` is the same as `!(a < b) && !(b < a)`

Reply:
Maybe try
```suggestion
    return (lhs._directed_edge & WeightedEdge::mask_undirected < rhs._directed_edge & WeightedEdge::mask_undirected);
```
This should fix the issue, and you won't need to write your own `operator==` and `makeUndirected()`. 
I don't think this would affect performance much, but could try.